File: CPP/No345.cpp
/**
 * Created by Xiaozhong on 2020/8/29.
 * Copyright (c) 2020/8/29 Xiaozhong. All rights reserved.
 */
#include "string"
#include "vector"
#include "iostream"


using namespace std;

class Solution {
public:
    string reverseVowels(string s) {
        string vowels = "aeiouAEIOU";
        int left = 0, right = s.size() - 1;
        while (left < right) {
            while (left < right && vowels.find(s[left]) == vowels.npos) left++;
            while (left < right && vowels.find(s[right]) == vowels.npos) right--;
            swap(s[left++], s[right--]);
        }
        return s;
    }
};

int main() {
    Solution s;
    cout << s.reverseVowels("hello");
}

File: Java/src/No345.java
/*
 * Copyright (c) 8/19/2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class No345 {
    public String reverseVowels(String s) {
        Set<Character> vowels = new HashSet<Character>(Arrays.asList('a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'));
        char[] chars = s.toCharArray();
        int n = s.length(), left = 0, right = n - 1;
        while (left < right) {
            while (left < right && !vowels.contains(chars[left])) left++;
            while (left < right && !vowels.contains(chars[right])) right--;
            if (left >= right) break;
            char t = chars[left];
            chars[left] = chars[right];
            chars[right] = t;
            left++;
            right--;
        }
        return new String(chars);
    }
}
